cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/ExecutionTaskManager.java

```diff
@@ -140,22 +199,102 @@ public void addBalancingProposals(Collection<BalancingProposal> proposals,
+      BalancingAction taskBalancingAction = tasks.iterator().next().proposal.balancingAction();
+      // Sanity check: All tasks must have the same balancing action.
+      for (ExecutionTask task : tasks) {
+        if (task.proposal.balancingAction() != taskBalancingAction) {
+          throw new IllegalStateException("Attempt to execute tasks with heterogeneous balancing actions.");
         }
-        if (task.destinationBrokerId() != null) {
-          _inProgressPartMovementsByBrokerId.put(task.destinationBrokerId(),
-                                                 _inProgressPartMovementsByBrokerId.get(task.destinationBrokerId()) + 1);
+      }
+
+      for (ExecutionTask task : tasks) {
+        _inProgressTasks.add(task);
+        _inProgressPartitions.add(task.proposal.topicPartition());
+        if (taskBalancingAction == BalancingAction.REPLICA_MOVEMENT) {
+          if (task.sourceBrokerId() != null) {
+            _inProgressPartMovementsByBrokerId.put(task.sourceBrokerId(),
+                                                   _inProgressPartMovementsByBrokerId.get(task.sourceBrokerId()) + 1);
+          }
+          if (task.destinationBrokerId() != null) {
+            _inProgressPartMovementsByBrokerId.put(task.destinationBrokerId(),
+                                                   _inProgressPartMovementsByBrokerId.get(task.destinationBrokerId()) + 1);
+          }
         }
       }
+
+      // Mark in progress and pending tasks.
+      if (taskBalancingAction == BalancingAction.REPLICA_MOVEMENT) {
+        _executionCounterByName.get(COUNTER_REPLICA_MOVE_IN_PROGRESS).inc(tasks.size());
+        _executionCounterByName.get(COUNTER_REPLICA_MOVE_PENDING).dec(tasks.size());
+      } else if (taskBalancingAction == BalancingAction.LEADERSHIP_MOVEMENT) {
+        _executionCounterByName.get(COUNTER_LEADERSHIP_MOVE_IN_PROGRESS).inc(tasks.size());
+        _executionCounterByName.get(COUNTER_LEADERSHIP_MOVE_PENDING).dec(tasks.size());
+      } else if (taskBalancingAction == BalancingAction.REPLICA_ADDITION) {
+        _executionCounterByName.get(COUNTER_REPLICA_ADDITION_IN_PROGRESS).inc(tasks.size());
+        _executionCounterByName.get(COUNTER_REPLICA_ADDITION_PENDING).dec(tasks.size());
+      } else if (taskBalancingAction == BalancingAction.REPLICA_DELETION) {
+        _executionCounterByName.get(COUNTER_REPLICA_DELETION_IN_PROGRESS).inc(tasks.size());
+        _executionCounterByName.get(COUNTER_REPLICA_DELETION_PENDING).dec(tasks.size());
+      } else {
+        throw new IllegalStateException(String.format("Unrecognized balancing action: %s.", taskBalancingAction));
+      }
+    }
+  }
+
+  /**
+   * Update relevant sensors upon topic deletion of an ongoing task based on the given task balancing action.
+   *
+   * @param taskBalancingAction Task balancing action for the ongoing task.
+   */
+  void updateSensorsUponTaskTopicDeletion(BalancingAction taskBalancingAction) {
```

Comment:
This method is a little weird. Should this be just markTaskDone() then markTaskAborted()?